BlockPreview: add __experimentalOnReady property

retrofox opened the pull request:
## Description

This PR adds a new property called **__experimentalOnReady** to the `<BlockPreview />` component, which allows us to know the exact moment when the preview is ready, providing useful data such as the scale factor applied, position and the reference to the DOM element.

It worths to mention that the whole process of previewing contemplates some tricks, pretty common when it needs to interact with the real DOM tree btw, such as adding a timeout (100ms), which sometimes makes it unpredictable and prone to fall in some race conditions if you aren't being careful.

Also, it adds the `__experimentalScalingDelay_` in order to allow set explicitly the delay to take before to start to compute the scale factor. The default value is `100ms`.

## How it works?

```es6
<BlockPreview
	blocks={ blocks }
	__experimentalOnReady={ ( { scale, previewContainerRef, position } ) => {
		console.log( `scale ${ scale } applied to the <${ previewContainerRef.current.tagName }> element.` );
		console.log( `at x: ${ position.x }, y: ${ position.y } position.` );
	} }
/>
```


## How has this been tested?

To test it you can tide a function to this property and confirm that it's called once the preview has been performed, for instance:

```es6
<BlockPreview onReady={ console.log } />
```

Chrome dev console:
![image](https://user-images.githubusercontent.com/77539/65524072-94538000-dec3-11e9-946c-7225d1b41932.png)


## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

retrofox committed:
block-preview: update doc

jorgefilipecosta commented on packages/block-editor/src/components/block-preview/README.md:
```diff
@@ -29,3 +29,17 @@ A block instance (object) or a blocks array you would like to render a preview.
+* **Type** `Function`
+* **Default:** `noop`
+
+Use this callback as an opportunity to know when the preview is ready. The callback will pass the scale factor, offsets position (x, y) and the DOM element reference. Eg:
```
Should we say:
will pass the scale factor, offsets position (x, y) and the DOM element reference, _if available_

The mobile application will also probably need to use this component and there we may not have a dom element for example.

jorgefilipecosta commented on packages/block-editor/src/components/block-preview/README.md:
````diff
@@ -29,3 +29,17 @@ A block instance (object) or a blocks array you would like to render a preview.
+```es6
+
+<BlockPreview
+    onReady={ ( { scale } ) => console.log( `Current preview scale: ${ scale }` )
````
nitpick: Should we use tabs instead of spaces?

jorgefilipecosta commented on packages/block-editor/src/components/block-preview/README.md:
```diff
@@ -29,3 +29,17 @@ A block instance (object) or a blocks array you would like to render a preview.
 
 `viewportWidth` can be used to simulate how blocks look on different device sizes or to make sure make sure multiple previews will be rendered with the same scale, regardless of their content.
+
+### `onReady`
```
The preview component suffered many changes recently. Would it be worth to make the prop experimental for two versions? Until we stabilize a little bit on its API.

retrofox replied:
Agree. let me update it to `__experimentalOnReady_.

jorgefilipecosta commented on packages/block-editor/src/components/block-preview/index.js:
```diff
@@ -55,9 +55,17 @@ function ScaledBlockPreview( { blocks, viewportWidth } ) {
-				setPreviewScale( containerElementRect.width / viewportWidth );
+				const scale = containerElementRect.width / viewportWidth;
+				setPreviewScale( scale );
+				onReady( { scale, ref: previewRef } );
```
I'm not sure if we should be passing refs via an event handler it seems to be uncommon, can't we use the normal way of passing references in React? e.g., in this case, BlockPreview may have an optional __experimental ScaledPreviewRef property that is passed to ScaledBlockPreview component which then forwards the ref.

retrofox replied:
I'm not understanding this feedback, Jorge :-/

> can't we use the normal way of passing references in React?

The reference is gotten just here, in this component, and the callback passes the reference as an argument as a chance to get keep working in the preview process if it's needed.

jorgefilipecosta replied:
Hi @retrofox I'm suggesting something identical to this commit https://github.com/WordPress/gutenberg/commit/e7aaee96f67eb7f5a27f16f8527f9f989003a92b where we allow the possibility of the BlockPreview user to pass its own reference, so the event does not need to pass it.

retrofox replied:
Sounds Good 👌 Let me update the PR.

retrofox replied:
Sorry, I think I completely misunderstood the conversation here:

> I'm not sure if we should be passing refs via an event handler

it isn't passing the reference. What is does is to give the reference as an argument in the callback function.

What the component does is adding a reference to be able to compute sizes in order to adjust the scale factor to apply to the preview block.

Removing the commit.

jorgefilipecosta reviewed (commented):
Hi @retrofox, thank you for your contribution things look good I just left some minor comments.
I guess onRead will be useful in some situations but I'm not imagining many possible use cases right now, could you elaborate on the concrete use case for onReady and why it is needed?

retrofox commented:
Thanks @jorgefilipecosta for your comments. Really appreciate it.

> could you elaborate on the concrete use case for onReady and why it is needed?

Specifically in our case what we need is to know the scale factor to apply it to other elements/components which are next to the blocks previewed with the `<BlockPreview />` component. The `<BlockPreview />` component processes the blocks to preview calling a function (setTimeout) asynchronously, which makes almost impossible to know when it's ready.

retrofox committed:
block-preview: set onReady as experimental prop

jorgefilipecosta commented on packages/block-editor/src/components/block-preview/index.js:
```diff
@@ -55,9 +55,16 @@ function ScaledBlockPreview( { blocks, viewportWidth } ) {
-				setPreviewScale( containerElementRect.width / viewportWidth );
+				const scale = containerElementRect.width / viewportWidth;
+				setPreviewScale( scale );
+				onReady( { scale, ref: previewRef } );
```
Now we don't need to pass the ref in the event anymore right?

retrofox replied:
Added a comment here (edited as well, sorry)
https://github.com/WordPress/gutenberg/pull/17242#discussion_r327164265

getdave commented on packages/block-editor/src/components/block-preview/index.js:
```diff
@@ -55,9 +55,17 @@ function ScaledBlockPreview( { blocks, viewportWidth } ) {
 			} else {
 				const containerElementRect = containerElement.getBoundingClientRect();
-				setPreviewScale( containerElementRect.width / viewportWidth );
+				const scale = containerElementRect.width / viewportWidth;
```
How come this code got added? Why was it added and what does it do? Doesn't seem directly connected to having a callback.

getdave reviewed (commented):
Thanks for working on this @retrofox! This looks go be a good idea. I had a similar thought when I was knee-deep in this component. Not being able to easily tell when the preview is "ready" was a major flaw.

Some minor tweaks here.

@jorgefilipecosta I feel like this entire component has become a lot more complex but we have no test coverage for it. Do you have concerns about this? If so how could one go about writing sane tests for this?

retrofox committed:
rename reference name

retrofox committed:
declare `position` constant

retrofox committed:
improve doc

retrofox committed:
use onReady name internally

retrofox committed:
pass the online styles to the callback

retrofox committed:
simply running onReady callback

retrofox committed:
avoid var name collisions

retrofox committed:
block-preview: reduce the computing delay to 0

retrofox committed:
block-preview: be able to set previewing delay

jorgefilipecosta commented on packages/block-editor/src/components/block-preview/index.js:
```diff
@@ -18,7 +18,15 @@ import BlockEditorProvider from '../provider';
 import { getBlockPreviewContainerDOMNode } from '../../utils/dom';
 
-function ScaledBlockPreview( { blocks, viewportWidth } ) {
+const getOnlineStyles = ( scale, x, y, isReady, width ) => ( {
```
getOnlineStyles = getInlineStyles?

retrofox replied:
🤦‍♂

jorgefilipecosta commented on packages/block-editor/src/components/block-preview/README.md:
```diff
@@ -29,3 +29,26 @@ A block instance (object) or a blocks array you would like to render a preview.
+* `scale`: the scale factor
+* `position`: offsets position (x, y)
+* `previewContainerRef`: DOM element reference
+* `styles`: Online styles applied to the preview container
```
Did you mean "Inline styles applied.."? 
Should the property be called inlineStyles instead of styles to be more specific?

jorgefilipecosta commented on packages/block-editor/src/components/block-preview/index.js:
```diff
@@ -98,7 +112,7 @@ function ScaledBlockPreview( { blocks, viewportWidth } ) {
 }
 
-export function BlockPreview( { blocks, viewportWidth = 700, settings } ) {
+export function BlockPreview( { blocks, viewportWidth = 700, settings, __experimentalOnReady = noop, __experimentalDelay = 100 } ) {
```
The delay we have is for the addition of scaling right? Maybe we can call the prop __experimentalScallingDelay to be more specific I guess ideally the name would refer to what is being delayed.
I think we should also document this prop together with __experimentalOnReady.

retrofox replied:
with only one `l` ;-)

jorgefilipecosta reviewed (commented):
Thank you for the iterations @retrofox. This PR seems mostly ready I just left some minor comments.

retrofox committed:
rename function/vars. Improve docs.

getdave commented on packages/block-editor/src/components/block-preview/README.md:
```diff
@@ -29,3 +29,32 @@ A block instance (object) or a blocks array you would like to render a preview.
+* **Type** `Int`
+* **Default** `100ms`
+
+It defines the delay before to start to calc the scale factor and position of the preview block.
```
How about

> Defines a delay to be applied before calculating the scale factor and position of the preview block.

jorgefilipecosta reviewed (approved):
Thank you for the iterations @retrofox, I tested the new props and they seem to work as expected. The code also looks good 👍

retrofox commented:
> Thank you for the iterations @retrofox, I tested the new props and they seem to work as expected. The code also looks good 👍

Thank you! really appreciate your help reviewing and developing the code. 🙇

jorgefilipecosta commented:
Hi @retrofox, was there an alternative to __experimentalOnReady found? If not, should the PR be rebased and merged?

retrofox commented:
> Hi @retrofox, was there an alternative to __experimentalOnReady found? If not, should the PR be rebased and merged?

Hi @jorgefilipecosta, I'm not aware of an alternative to this approach. PR is already rebased.

retrofox committed:
block-preview: improve Readme doc

retrofox committed:
block-preview: fix x position. simplify.

brentswisher committed:
Docs: Fix JavaScript linting command (#20371)

The docs say to use run `npm run lint:fix` to automatically fix some issues, that doesn't exist, it should be `npm run lint-js:fix`
